Feature : added new PlanarControls

EmmanuelSchmuck opened the pull request:
PlanarControls provides ergonomic camera controls for the planar mode : drag, pan, orbit,
zoom and animated expression to smoothly move and orient the camera.

Modified planar example to use these controls
Update contributors

EmmanuelSchmuck committed:
Feature : added new PlanarControls

PlanarControls offers ergonomic camera controls for the planar mode : drag, pan, orbit,
zoom and animated expression to smoothly move and orient the camera.

Modified planar example to use these controls
Update contributors

Modifications following peppsac review :
-all external variables are now within the class, this allows to
instanciate more than one controller.
-clamp01 function removed, using THREE clamp instead.
-various style modifications (separator removed, const array format...)

peppsac commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,762 @@
+// time management
+let deltaTime = 0;
+let lastElapsedTime = 0;
+const clock = new THREE.Clock();
```
How about storing all these inside the `PlanarControls` instance so a user can instanciate several controls?

peppsac commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,762 @@
+        dragDelta.set(0, 0, 0);
+    };
+
+    /** =========================================================================================================================
```
I don't mind the `=========` separator but this is something that should be done everywhere or nowhere (as all other stylistic choices).
So I'd suggest removing them from this PR and open a specific PR to convert itowns to this style (but I'm not sure it would be merged...)

peppsac commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,762 @@
+* @param {number} value : the value to be smoothed, between 0 and 1
+* @returns {number}
+*/
+var clamp01 = function clamp01(value) {
```
You should directly use `THREE.Math.clamp(value, 0, 1)` instead of adding a one liner function.

peppsac commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,762 @@
+
+// event keycode
+const keys = { CTRL: 17, SPACE: 32, S: 83, T: 84 };
+const mouseButtons = { LEFTCLICK: THREE.MOUSE.LEFT, MIDDLECLICK: THREE.MOUSE.MIDDLE, RIGHTCLICK: THREE.MOUSE.RIGHT };
```
Minor style comment:
```js
 const mouseButtons = {
    LEFTCLICK: THREE.MOUSE.LEFT,
    MIDDLECLICK: THREE.MOUSE.MIDDLE,
    RIGHTCLICK: THREE.MOUSE.RIGHT,
};
```
is prefered.

peppsac reviewed (commented):
Thanks for the PR and welcome to iTowns project :)

I did a very quick first review.
I (or another reviewer) will have do a more in depth one before merging but the code is looking good mostly and is well documented, nice!

EmmanuelSchmuck commented:
Thank you for the review @peppsac , I made the modifications you suggested.

peppsac commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,746 @@
+    this.view = view;
+    this.camera = view.camera.camera3D;
+    this.domElement = view.mainLoop.gfxEngine.renderer.domElement;
+    this.position = this.camera.position;
```
`this.position` isn't really useful: let's use `this.camera.position` directly instead.

peppsac commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,746 @@
+    * Updates the view and camera if needed, and handles the animated travel
+    */
+    this.update = function update() {
+        this.deltaTime = this.clock.getElapsedTime() - this.lastElapsedTime;
```
Why don't you use the `dt` provided as an argument to your update function (see https://github.com/iTowns/itowns/blob/master/src/Core/MainLoop.js#L63)?

peppsac commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,746 @@
+            this.handleTravel(this.deltaTime);
+        }
+        if (this.state !== STATE.NONE) {
+            this.view.camera.update(window.innerWidth, window.innerHeight);
```
You don't need to call this: it's done for you in view's update process (https://github.com/iTowns/itowns/blob/master/src/Core/MainLoop.js#L96)

peppsac reviewed (commented):
I tested this controls quickly and it seems to be working good.

I left a few comments and then it'll be ready for merging.

EmmanuelSchmuck committed:
planarcontrols modifications following peppsac review :
update() uses mainloop's deltatime, no longer call camera.update()
this.camera.position replaces this.position

EmmanuelSchmuck commented:
@peppsac : modified again following your suggestions. Also, my inputs no longer call update() but notifyChange() instead, much cleaner.

EmmanuelSchmuck committed:
removed options.startposition and startlook, start view with Y instead of S (used for picking)

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+    // time management
+    this.deltaTime = 0;
+    this.lastElapsedTime = 0;
+    this.clock = new THREE.Clock();
```
`this.clock` is not used.

autra replied:
Actually, all three "time management" vars are not used, if I'm not mistaken.

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+    this._handlerOnMouseDown = onMouseDown.bind(this);
+    this._handlerOnMouseUp = onMouseUp.bind(this);
+    this._handlerOnMouseMove = onMouseMove.bind(this);
+    this._handlerOnMouseWheel = onMouseWheel.bind(this);
```
As everything is defined in the constructor, you don't really need to attach these handlers on `this`. They can be local variables. That's better because they are truly isolated then (and being able to really isolate things is a very good reason to do this way, and not to use `prototype` or ES6 classes :-) )

autra replied:
This is also true for every local state you don't want to expose. I'm thinking about:
- this.view
- this.camera
- all `travelFoo` vars
- all `drag` vars
-  this.mousePosition, this.lastMousePosition and this.deltaMousePosition
- isCtrlDown

I'm pretty sure these should be hidden.

In contrast, I believe keeping all the option properties visible is perfect :-)

For `this.state`, I have no strong opinion about it.

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+    * Updates the view and camera if needed, and handles the animated travel
+    * @param {number} dt : deltatime given by mainLoop
+    * @param {boolean} updateLoopRestarted : given by mainLoop
+    */
```
Not sure we want to document this. It's just the implementation of a FrameRequester. It's not meant to be called by the outside world. Better not have it in documentation.

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+            dt = 16;
+        }
+        if (this.state === STATE.TRAVEL) {
+            this.handleTravel(dt / 1000);
```
What about reasoning in ms everywhere?

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+* @returns {number}
+*/
+var smooth = function smooth(value) {
+    // p between 1.0 and 1.5
```
p is 1.20?

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+    }
+    if (event.keyCode === keys.CTRL) {
+        this.isCtrlDown = true;
+    }
```
nits: switch-case?

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+*/
+var onKeyUp = function onKeyUp(event) {
+    if (event.keyCode === keys.CTRL) {
+        this.isCtrlDown = false;
```
I believe you don't need `isCtrlDown`, use event.ctrlKey instead (available for [MouseEvent](https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/ctrlKey) and [KeyboardEvent](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/ctrlKey)).

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+        this.view.notifyChange(true);
+    }
+    if (this.state === STATE.ROTATE) {
+        this.handleRotation();
```
So we call `handleRotation` at each mouse event. I have the feeling that doing all this work in `update` instead would yield the same result with better perf (actually that's one of the reason I wrote the FrameRequester logic : to give a painless mechanism to throttle input events). The logic is to set the state in event handlers, then deal with them (movements, animation etc... in `update`). 

I know `FlyControls` doesn't do things this way, but it should have done so imho (my bad). Or maybe I missed something?

autra replied:
To elaborate: sometimes you change the position of the camera in an event (ex `handleRotation`), sometimes you change it in `update` (ex the drag logic). I believe that we should do calculations in the `update` macrotask (meaning: in the same macrotask of the `requestAnimationFrame`) as much as possible, not in the events macrotasks. So here, you'd only need to record positions.

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+    this.lastMousePosition.copy(this.mousePosition);
+
+    if (event.button === mouseButtons.LEFTCLICK) {
+        if (this.isCtrlDown) {
```
Use `event.ctrlKey` instead (see below).

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+        }
+        else if (this.state === STATE.ROTATE) {
+            this.domElement.style.cursor = 'move';
+        }
```
switch-case block?

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+    * removes all the input event listeners (desactivate the controls)
+    */
+    this.removeInputListeners = function removeInputListeners() {
+        window.removeEventListener('keydown', this._handlerOnKeyDown, true);
```
I agree this looks more practical in our examples, but please consider that one html page could have several itowns instance, or other js actions using keys... So I'll listen to `domElement` events instead.

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+    * @returns {THREE.Vector3}
+    */
+    this.getWorldPointFromMathPlaneAtScreenXY = function getWorldPointFromMathPlaneAtScreenXY(posXY, altitude) {
+        const vector = new THREE.Vector3();
```
Let's put this temp var in a closure to avoid instanciating it at each frame (by experience, creating these temp vars is very costly and the cause of every movement stuttering I had when implementing controls). We only need one per functions.

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+    */
+    this.getWorldPointAtScreenXY = function getWorldPointAtScreenXY(posXY) {
+        // the returned value
+        const pointUnderCursor = new THREE.Vector3();
```
Put this temp var in a closure.

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+    * @param {number} altitude : the altitude (z) of the mathematical plane
+    * @returns {THREE.Vector3}
+    */
+    this.getWorldPointFromMathPlaneAtScreenXY = function getWorldPointFromMathPlaneAtScreenXY(posXY, altitude) {
```
I have the feeling this might be better in PlanarView (same thing for `getWorldPointAtScreenXY`). I'm not sure about that though. WDYT?

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+
+    /**
+    * Handle the drag movement (translation on xy plane) when user moves the mouse while in STATE.DRAG
+    * The drag movement is previously initiated when user does a left-click, by initiateDrag()
```
> The drag movement is previously initiated when user does a left-click, by initiateDrag()

Maybe replace by

> The drag movement is previously initiated by initiateDrag()

because the fact that drag movement is done by left-click is an ergonomy detail (and could be made configurable in the future).

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+        const panSpeed = THREE.Math.lerp(this.minPanSpeed, this.maxPanSpeed, distToGround);
+
+        // lateral movement (local x axis)
+        this.camera.position.copy(this.camera.localToWorld(new THREE.Vector3(panSpeed * -1 * this.deltaMousePosition.x, 0, 0)));
```
Use `this.camera.position.set(x, y, z)` directly.

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+
+        const screenCenter = new THREE.Vector2(0.5 * window.innerWidth, 0.5 * window.innerHeight);
+
+        this.centerPoint.copy(this.getWorldPointAtScreenXY(screenCenter));
```
You could pass directly `{x: 0.5 * window.innerWidth, y: 0.5 * window.innerHeight}` to `getWorldPointAtScreenXY`, I *think* it'll be cheaper than to create a Vector2... or modify getWorldPointAtScreenXY to take directly x and y? Caller would not need to create an object this way.

autra replied:
(and you need to use canvas.clientHeight)

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+    this.initiateRotation = function initiateRotation() {
+        this.state = STATE.ROTATE;
+
+        const screenCenter = new THREE.Vector2(0.5 * window.innerWidth, 0.5 * window.innerHeight);
```
You want the center of the canvas, not the window (remember, the canvas can be integrated in whatever webpage). You already have a ref to the canvas (`domElement`) and you'll need to use `clientHeight` and `clientWidth` (I guess). Please also change other instances of windows.innerWidth in this file.

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+    this.handleRotation = function handleRotation() {
+        // angle deltas
+        // deltaMousePosition is computed in onMouseMove / onMouseDown s
+        const thetaDelta = -this.rotateSpeed * this.deltaMousePosition.x / window.innerWidth;
```
same, I'm pretty sure you want the canvas clientWidth.

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+        // the vector from centerPoint (focus point) to camera position
+        const offset = this.camera.position.clone().sub(this.centerPoint);
+
+        const quat = new THREE.Quaternion().setFromUnitVectors(this.camera.up, new THREE.Vector3(0, 0, 1));
```
Create the quaternion in a closure.

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+    */
+    this.initiateSmartZoom = function initiateSmartZoom() {
+        // point under mouse cursor
+        const pointUnderCursor = this.getWorldPointAtScreenXY(this.mousePosition);
```
Should we limit the smartzoom action to a click in the plane limits?

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+        const distanceToPoint = this.camera.position.distanceTo(pointUnderCursor);
+
+        // camera height (altitude above ground) at the end of the travel
+        const targetHeight = THREE.Math.lerp(this.smartZoomHeightMin, this.smartZoomHeightMax, Math.min(distanceToPoint / 5000, 1));
```
5000 seems to be a magic number... Beside, I would have expected something involving only z, not all components of the distance, can you elaborate on this calculus?

autra replied:
Or seeing what happens below, maybe `targetHeight` could be named differently?

Btw, "I found it nicer this way" is a perfect valid answer here :-)

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+        // setup the end rotation :
+
+        // case where targetOrientation is a quaternion
+        if (typeof targetOrientation.w !== 'undefined') {
```
Then the test should be `if (targetOrientation instanceof THREE.Quaternion)`?

autra replied:
(it would also work if `targetOrientation` is undefined this way)

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+        // update cursor
+        this.updateMouseCursorType();
+
+        this.travelUseRotation = !(targetOrientation === 'none');
```
I would test `targetOrientation instanceOf THREE.Quaternion || targetOrientation instanceOf THREE.Vector3`, it would also cover `undefined`, `null`, whatever string and invalid objects the caller gave you. Here the caller needs to remember to pass `'none'` to explicitly disable rotation.

Not using `isVector3` because this would throw an error if undefined or null. (I don't know why Vector3 implementors felt the need to have this boolean, it's useless imho).

autra replied:
(I assume you don't want a rotation to the default value of `travelEndRot` if `targetOrientation` is invalid)

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+        if (this.instantTravel) {
+            this.travelDuration = 0;
+        }
+
```
nits: remove empty line (a reader could miss the `else if` below)

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+        // case where travelTime is set to 'auto' : travelDuration will be a value between autoTravelTimeMin and autoTravelTimeMax
+        // depending on travel distance and travel angular difference
+        else if (travelTime === 'auto') {
+            // a value between 0 and 1 according to the travel distance. Adjusted by autoTravelTimeDist parameter
```
then `autoTravelTimeDist` is a factor, not a real distance? Maybe we could rename it then.

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,736 @@
+        const screenCenter = new THREE.Vector2(0.5 * window.innerWidth, 0.5 * window.innerHeight);
+
+        topViewPos.copy(this.getWorldPointAtScreenXY(screenCenter));
+        topViewPos.z += Math.min(this.maxAltitude, this.camera.position.distanceTo(topViewPos));
```
Can you explain this line to me? Why isn't it always `maxAltitude`?

autra reviewed (changes requested):
Taking over this, as @peppsac is off. Sorry for the inconvenience to have another reviewer with different expectations stepping in :-/

On overall I really like it! The ergonomy (and the code) is good :-) I made a lot of comments, but most are minor.

EmmanuelSchmuck committed:
many modifications following review by Autra, see PR

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -505,16 +526,20 @@ function PlanarControls(view, options = {}) {
-        return pointUnderCursor;
-    };
+        let dir = new THREE.Vector3();
+        let pointUnderCursor = new THREE.Vector3();
```
It's not a big deal but, `dir` does not need to be declared here. You can declare it when you first use it. Anyway, `dir` and `vector` will store the same pointer, because `unproject`, `sub`, `normalize` and `multiplyScalar` all modify the vector3 inplace and returns it (see for instance [how `sub` is implemented](https://github.com/mrdoob/three.js/blob/master/src/math/Vector3.js#L161)).

On the other hand, `pointUnderCursor` **must not** be declared here, otherwise we would always return the same pointer, which is dangerous. We need to create a new instance each time, which is done by the `clone` anyway. So in short, just
```js
return this.camera.clone().add(dir.multiplyScalar(distance));
``` 
instead of line 538-540 is fine (that being said, you can keep `pointUnderCursor` if you find it clearer, just declare it with const on line 538).

autra replied:
And it's totally my fault, because I was the one to ask you about `pointUnderCursor`. Sorry :-/

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -523,67 +548,78 @@ function PlanarControls(view, options = {}) {
         const pointUnderCursor = new THREE.Vector3();
+        return (posXY) => {
+            // check if there is valid geometry under cursor
+            if (typeof this.view.getPickingPositionFromDepth(posXY) !== 'undefined') {
```
Sorry I missed that in my previous review: let's put `this.view.getPickingPositionFromDepth(posXY)` in a variable to avoid calling it 2 times.

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -212,10 +223,10 @@ function PlanarControls(view, options = {}) {
 
         // lateral movement (local x axis)
-        this.camera.position.copy(this.camera.localToWorld(new THREE.Vector3(panSpeed * -1 * this.deltaMousePosition.x, 0, 0)));
+        this.camera.position.copy(this.camera.localToWorld(new THREE.Vector3(panSpeed * -1 * deltaMousePosition.x, 0, 0)));
```
```js
this.handlePanMovement  = (() => {
    const tmp = new THREE.Vector3();
    
    return () => {
         // ....
         tmp.set(panSpeed * -1 * deltaMousePosition.x, 0, 0);
         this.camera.position.copy(this.camera.localToWorld(tmp));
         // ...
   }
})();
```
should prevent one instanciation.

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -243,53 +252,58 @@ function PlanarControls(view, options = {}) {
+            const offset = this.camera.position.clone().sub(centerPoint);
+
+            quat.setFromUnitVectors(this.camera.up, new THREE.Vector3(0, 0, 1));
+            quatInverse = quat.clone().inverse();
```
what about `quat.setFromUnitVectors(this.camera.up, new THREE.Vector3(0, 0, 1)).inverse()` and using `quat` instead of quatInverse below? It saves a `clone()`.

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -243,53 +252,58 @@ function PlanarControls(view, options = {}) {
-                offset.applyQuaternion(rotationZQuaternion);
+    this.handleRotation = (() => {
+        const quat = new THREE.Quaternion();
+        let quatInverse = new THREE.Quaternion();
```
Please "cache" here:

- `new THREE.Vector3(0, 0, 1)` on line 268
- `rotationXQuaternion` and `vector` on line 279-280
- `rotationZQuaternion` on line 290

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -523,67 +548,78 @@ function PlanarControls(view, options = {}) {
+            else {
+                pointUnderCursor.copy(this.getWorldPointFromMathPlaneAtScreenXY(posXY, this.groundLevel));
+            }
+            return pointUnderCursor;
```
same thing as above. Sorry, my fault :-/

autra commented on src/Renderer/ThreeExtended/PlanarControls.js:
```diff
@@ -0,0 +1,743 @@
+    };
+
+    PlanarControls.prototype = Object.create(THREE.EventDispatcher.prototype);
+    PlanarControls.prototype.constructor = PlanarControls;
```
You're not using `dispatchEvent`, so you don't need the 2 lines above.

autra reviewed (changes requested):
Nearly there :-)

EmmanuelSchmuck committed:
additional changes following @autra second review (opti, cleanup)

autra commented:
Yay thanks \o/

autra commented:
Nice first contrib btw !

EmmanuelSchmuck commented:
Thank you :)
